src/credentials/CHIPOperationalCredentials.h

```diff
@@ -0,0 +1,236 @@
+     * @param cert              A pointer to the valid CHIP certificate that matches search criteria.
+     *
+     * @return Returns a CHIP_ERROR if no valid certificate could be found
+     **/
+    CHIP_ERROR FindValidCert(const CertificateKeyId & trustedRootId, const ChipDN & subjectDN,
+                             const CertificateKeyId & subjectKeyId, ValidationContext & context, ChipCertificateData *& cert);
+
+    /**
+     * @brief A function to sign a msg using ECDSA and the respective device credentials keypair.
+     *
+     * @param msg           Message that needs to be signed
+     * @param msg_length    Length of message
+     * @param out_signature Buffer that will hold the output signature. The signature consists of: 2 EC elements (r and s),
+     *                      represented as ASN.1 DER integers, plus the ASN.1 sequence Header
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR SignMsg(const CertificateKeyId & trustedRootId, const uint8_t * msg, const size_t msg_length,
+                       P256ECDSASignature & out_signature);
+
+    /**
+     * @return A pointer to device credentials (in x509 format).
+     **/
+    const uint8_t * GetDevOpCred(const CertificateKeyId & trustedRootId) const
+    {
+        return mChipDeviceCredentials.at(trustedRootId).mCredential;
+    }
+
+    /**
+     * @return Length of the loaded device credentials buffer.
+     **/
+    uint16_t GetDevOpCredLen(const CertificateKeyId & trustedRootId) const { return mChipDeviceCredentials.at(trustedRootId).mLen; }
+
+    CHIP_ERROR SetDevOpCred(const CertificateKeyId & trustedRootId, const uint8_t * chipDeviceCredentials,
+                            uint16_t chipDeviceCredentialsLen);
+
+    const P256Keypair & GetDevOpCredKeypair(const CertificateKeyId & trustedRootId) const
+    {
+        return *mDeviceOpCredKeypair.at(trustedRootId);
+    }
+
+    CHIP_ERROR SetDevOpCredKeypair(const CertificateKeyId & trustedRootId, P256Keypair * newKeypair);
+
+private:
+    ChipCertificateSet * mOpCreds; /**< Pointer to an array of certificate data. */
+    uint8_t mOpCredCount;          /**< Number of certificates in mOpCreds
+                                    array. We maintain the invariant that all
+                                    the slots at indices less than
+                                    mCertCount have been constructed and slots
+                                    at indices >= mCertCount have either never
+                                    had their constructor called, or have had
+                                    their destructor called since then. */
+    uint8_t mMaxCerts;             /**< Length of mOpCreds array. */
+    bool mMemoryAllocInternal;     /**< Indicates whether temporary memory buffers are allocated internally. */
+    std::map<CertificateKeyId, NodeCredential, CertificateKeyIdCompare> mChipDeviceCredentials;
```

Comment:
std::map uses heap. Can we use some alternative?

Reply:
Agreed. Thanks, updated.